nextest-filtering/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+    PackageId,
+};
+
+pub mod error;
+mod parsing;
+
+use parsing::{Expr, RawNameMatcher, SetDef};
+
+/// Matcher for name
+///
+/// Used both for package name and test name
+#[derive(Debug, Clone)]
+pub enum NameMatcher {
+    /// Exact value
+    Equal(String),
+    /// Simple contains test
+    Contains(String),
+    /// Test against a regex
+    Regex(regex::Regex),
+}
+
+impl PartialEq for NameMatcher {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Contains(s1), Self::Contains(s2)) => s1 == s2,
+            (Self::Equal(s1), Self::Equal(s2)) => s1 == s2,
+            (Self::Regex(r1), Self::Regex(r2)) => r1.as_str() == r2.as_str(),
+            _ => false,
+        }
+    }
+}
+
+impl Eq for NameMatcher {}
+
+/// Define a set of tests
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum FilteringSet {
+    /// All tests in packages
+    Packages(HashSet<PackageId>),
+    /// All tests matching a name
+    Test(NameMatcher),
+    /// All tests
+    All,
+    /// No tests
+    None,
+    // Possible addition: Binary(NameMatcher)
+}
+
+/// Filtering expression
+///
+/// Used to filter tests to run.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum FilteringExpr {
+    /// Accepts every tests not in the given expression
+    Not(Box<FilteringExpr>),
+    /// Accepts every tests in either given expression
+    Union(Box<FilteringExpr>, Box<FilteringExpr>),
+    /// Accepts every tests in both given expression
+    Intersection(Box<FilteringExpr>, Box<FilteringExpr>),
+    /// Accepts every tests in a set
+    Set(FilteringSet),
+}
+
+impl NameMatcher {
+    pub fn is_match(&self, input: &str) -> bool {
+        match self {
+            Self::Equal(text) => text == input,
+            Self::Contains(text) => input.contains(text),
+            Self::Regex(reg) => reg.is_match(input),
+        }
+    }
+}
+
+impl FilteringSet {
+    fn includes(&self, package_id: &PackageId, name: &str) -> bool {
+        match self {
+            Self::All => true,
+            Self::None => false,
+            Self::Test(matcher) => matcher.is_match(name),
+            Self::Packages(packages) => packages.contains(package_id),
+        }
+    }
+}
+
+fn matching_packages(
+    matcher: &NameMatcher,
+    all_packages: &[PackageMetadata],
+) -> HashSet<PackageId> {
+    all_packages
+        .iter()
+        .filter(|p| matcher.is_match(p.name()))
+        .map(|p| p.id().clone())
+        .collect()
+}
+
+fn dependencies_packages(
+    matcher: &NameMatcher,
+    all_packages: &[PackageMetadata],
+    cache: &mut DependsCache,
+) -> HashSet<PackageId> {
+    let packages = all_packages
+        .iter()
+        .filter(|p| matcher.is_match(p.name()))
+        .map(|p| p.id());
+    let mut set = HashSet::new();
+    for id1 in packages {
+        for p2 in all_packages {
+            let id2 = p2.id();
+            if id1 == id2 {
+                continue;
+            }
```

Comment:
I'd match the behavior of https://bazel.build/reference/query here and in `rdependencies_packages` below and remove this check. `deps()` and `rdeps()` should return the transitive closure.

For a stretch goal, you could implement the depth limit listed there.